### ISSUE TITLE
Cannot grant accessibility access after updating to 2.2.1 / 3.0.0 Beta 4 [bug]

### ISSUE THREAD

**Issue by R3IDV3** [noah-nuebling/Mac-Mouse-Fix#412]
### __Update 3 by Noah__

In addition to 2.2.2, [3.0.0 Beta 5](https://github.com/noah-nuebling/mac-mouse-fix/releases/tag/3.0.0-Beta-5) now also implements the automatic fix to this problem!

The automatic fix won't work in some situations like on some older macOS versions. For those situations I've updated the [Accessibility Guide](https://github.com/noah-nuebling/mac-mouse-fix/discussions/101) to include manual steps for solving this problem.

I'll close this Issue now. If you still experience these problems, please open a new issue or let me know so I can re-open this one. Thank you!

### __Update 2 by Noah__

This issue should be **resolved** in the latest release [2.2.2](https://github.com/noah-nuebling/mac-mouse-fix/releases/tag/2.2.2)!

If you're on the **MMF 3 Beta**, a fix will be coming with Beta 5. Until then refer to [@rishifter's comment](https://github.com/noah-nuebling/mac-mouse-fix/issues/412#issuecomment-1285629403).

**Why was this happening?**

Here's my best understanding: I finally bought an Apple Developer program membership for $99 for MMF 3. (Which also means I will notarize the app at some point soon, which should prevent the "Malware" warnings when opening Mac Mouse Fix for the first time!). This membership allows you to do better "codesigning" for the app. Mac Mouse Fix 2.2.1 shipped with this new codesigning, but 2.2.0 didn't. So when you upgraded from 2.2.0 to 2.2.1 macOS got really confused. It seems that the security layer in macOS sees 2.2.1 and 2.2.0 as different apps due to the different signing so the accessibility permissions you gave to 2.2.0 don't transfer over to 2.2.1. But the System Settings **do** see 2.2.0 and 2.2.1 as the same app so the settings don't update. The result is that the toggle in the Accessibility Settings will still apply to MMF 2.2.0 even after you updated to 2.2.1. And so you can't grant accessibility access to 2.2.1.

The way to fix this is to remove the old 2.2.0 `Mac Mouse Fix Helper` from the security settings first and then restart the 2.2.1 `Mac Mouse Fix Helper` app so it can add itself to the Accessibility Settings properly.

2.2.2 and 3.0.0 Beta 5 will fix the problem by automatically doing these steps for you when the issue occurs.

### __Update by Noah__

Hey guys! I've looked into this all day and while I'm still not totally sure what causes it, I think I've found a way to __automatically fix it__.

It's implemented here:
[MacMouseFixApp-2.2.2-Alpha.zip](https://github.com/noah-nuebling/mac-mouse-fix/files/9843007/MacMouseFixApp-2.2.2-Alpha.zip)

Let me know if that works for you!

Otherwise try what @rishifter described [here](https://github.com/noah-nuebling/mac-mouse-fix/issues/412#issuecomment-1285629403).

---

Here's @R3IDV3 original issue:

## Description

After updating to 3.0.0 Beta 4, I was asked to again grant accessibility access. I followed the link and checked the box to enable the helper in security preferences, although this did not result in the MMF responding. MMF was then stuck and unable to be quit unless force quitted. Force quitting and reattempting did not fix this issue. To fix the issue, I had to remove the helper from the list in system preferences, then restart MMF and then enable the helper again in system preferences. FYI. I know system pref helper can be weird sometimes lol! Maybe there is a way to delete the old helper while installing and relaunching the updated version to avoid this issue? Loving the app!!

## Steps to reproduce

–

## Context

Mac Mouse Fix Version: 3.0.0 Beta 4
macOS Version: 12.3
Mouse Model: –

## Additional Comments

–

---

### Console Logs

–

### Crash Reports

–

**Comment by noah-nuebling** [comment]
Hey thanks for your feedback! 

I assume you upgraded from 3.0.0 Beta 3 right? 

Mac Mouse Fix does a bunch of things automatically to prevent interference with old versions still running in the background, but a bunch of those mechanisms are broken for **Beta 3**. This is because Apple forced me to change the apps identifier for Beta 3 for some reason which means that other versions don't recognize that Beta 3 is a version of Mac Mouse Fix and won't apply the custom code to prevent interference between the different versions.

**Beta 4** however is back to the proper identifier. So I hope that these issues will not occur anymore when you update to beta 5 or any versions after that!

**Comment by noah-nuebling** [comment]
Will leave this open for now for people with the same issue.

**Comment by noah-nuebling** [comment]
> Loving the app!!

Oh thanks for letting me know!! :)

**Comment by trivita** [comment]
i'm having this issue too, remove the app completely and then do a clean install didnt help (didnt reboot but no helper in process list).
waiting for beta 5

**Comment by noah-nuebling** [comment]
@trivita Oh you do have to reboot in my experience! Also make sure to empty the trash after deleting all versions of MMF and only then restart the computer and reinstall.

If you don't empty the trash macOS will keep trying to open the version of Mac Mouse Fix that's in the trash in my experience. Even after a reboot. And if you empty the trash but you don't reboot then macOS will still try to open the version of Mac Mouse Fix you just deleted even though it doesn't exist on your computer anymore. 

Edit: 

Just noticed I was assuming you're having another issue where Mac Mouse Fix **won't enable / won't enable properly**.
For that issue the solution steps are:

1. Disable Mac Mouse Fix and remove all versions of Mac Mouse Fix from your computer
2. Empty the Trash
3. Restart your computer
4. Reinstall Mac Mouse Fix 3.0.0 Beta 4

But if your problem is with **granting accessibility permissions** specifically, I don't have better solution steps than what @rishifter has provided.

**Comment by rishifter** [comment]
Sharing my experience in case it's helpful to others. I didn't have to reboot or reinstall. I've had similar issues with some apps earlier for permissions. I had to remove the app from the preferences (not just disable and enable them). Force quit MMF after and when launching it again it adds the appropriate helper in the pref for me to enable. Then it immediately accepts.

**Comment by R3IDV3** [comment]
> Sharing my experience in case it's helpful to others. I didn't have to reboot or reinstall. I've had similar issues with some apps earlier for permissions. I had to remove the app from the preferences (not just disable and enable them). Force quit MMF after and when launching it again it adds the appropriate helper in the pref for me to enable. Then it immediately accepts.

This is exactly what worked for me too

**Comment by trivita** [comment]
> Sharing my experience in case it's helpful to others. I didn't have to reboot or reinstall. I've had similar issues with some apps earlier for permissions. I had to remove the app from the preferences (not just disable and enable them). Force quit MMF after and when launching it again it adds the appropriate helper in the pref for me to enable. Then it immediately accepts.

yup, i figured out this method too.
despite this bug, i hope that more mice are supported (extra buttons recognized correctly)

**Comment by JensMertelmeyer** [comment]
I just had the same problem, but with the regular version, (Version 2.2.1 (7310)), I've never opted in for beta.

**Comment by DzigaVertoff** [comment]
> I just had the same problem, but with the regular version, (Version 2.2.1 (7310)), I've never opted in for beta.

Same here

**Comment by noah-nuebling** [comment]
Hey guys I've edited the original post with an update

**Comment by noah-nuebling** [comment]
Updated the original post again. Problem should be resolved in the latest update!

Let me know if you have more questions or issues!

**Comment by noah-nuebling** [comment]
Reopening until this problem is fixed in the 3.0.0 Beta as well

**Comment by noah-nuebling** [comment]
Problem fixed in 3.0.0 Beta 5. See update at the very top.
